app/forms/delivery_partner_form.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+class DeliveryPartnerForm
+  include ActiveModel::Model
+
+  attr_accessor :name, :lead_providers, :provider_relationships
+  validate :lead_providers_and_cohorts_validation
+
+  def available_lead_providers
+    LeadProvider.includes(:cohorts).select { |lead_provider| lead_provider.cohorts.any? }
+  end
+
+  def chosen_provider_relationships
+    provider_relationships.map do |relationship|
+      ProviderRelationship.new(
+        cohort: Cohort.find(relationship["cohort_id"]),
+        lead_provider: LeadProvider.find(relationship["lead_provider_id"]),
+      )
+    end
+  end
+
+  def display_lead_provider_details
+    chosen_provider_relationships.group_by(&:lead_provider).map do |lps_to_relationships|
+      {
+        name: lps_to_relationships.first.name,
+        chosen_cohorts: lps_to_relationships.second.map { |relationship| relationship.cohort.display_name }.join(", "),
+      }
+    end
+  end
+
+  def populate_provider_relationships(params)
+    self.lead_providers = []
+    self.provider_relationships = []
+
+    lead_provider_ids = params.dig(:delivery_partner_form, :lead_providers)&.keep_if(&:present?)
```

Comment:
I hope that if we make a form out of complete list of parameters, we won't have to do the digging in here.

Reply:
I don't understand what you mean

Reply:
I think the reason `populate_provider_relationships` needs to take `params` (and do the digging for content in them) is that in the controller method initialise the form with the old session parameters:
```
    @delivery_partner_form = DeliveryPartnerForm.new(session[:delivery_partner_form])
    @delivery_partner_form.populate_provider_relationships(params)

    unless @delivery_partner_form.valid?
      render :choose_lead_providers and return
    end

    session[:delivery_partner_form].merge!(
      {
        lead_providers: @delivery_partner_form.lead_providers,
        provider_relationships: @delivery_partner_form.provider_relationships,
      },
    )
```
If we do something like
```
    new_session_params = session[:delivery_partner_form].merge(params[:delivery_partner_form])
    @delivery_partner_form = DeliveryPartnerForm.new(new_session_params)
    @delivery_partner_form.populate_provider_relationships(params)

    unless @delivery_partner_form.valid?
      render :choose_lead_providers and return
    end

    session[:delivery_partner_form] = new_session_params
```
then perhaps at the time we call `populate_provider_relationships` we won't need to load lead providers ourselves - because hopefully the form will do it for us.

Reply:
I'm happy for this to be explored in a separate PR with less fluff, but I'd like to try it - because I think it will make the form shorter.

Reply:
Will address in separate PR 👍